File: python/polygon.py
from point import Point

class Polygon:
    #Construtor padrão
    def __init__(self, p, q, r):
        self.p1 = p
        self.p2 = q
        self.p3 = r

    def verifyColision(self, ray):
        v1 = self.p2 - self.p1
        v2 = self.p3 - self.p1
        norma = v1.crossProduct(v2)
        if ray.d.dotProduct(norma) == 0:
            return NULL
        tint = (p1 - R.p0).dotProduct(norma) / R.d.dotProduct(norma)
        Pint = R.pointT(tint);
        if ((p2-p1).crossProduct(Pint - p1)).dotProduct(norma) < 0 or ((p3-p2).crossProduct(Pint - p2)).dotProduct(norma) < 0 or ((p1-p3).crossProduct(Pint - p3)).dotProduct(norma) < 0:
            return NULL
        return tint

File: primitivos/Polygon.cpp
#include "Polygon.hpp"

Polygon::Polygon(Point p, Point q, Point r){
    p1 = p;
    p2 = q;
    p3 = r;
}
double Polygon::colisionPoints(Ray R){
    Vector v1 = Vector(p1, p2);
    Vector v2 = Vector(p1, p3);
    Vector norma = v1.crossProduct(v2);
    if(R.d.dotProduct(norma) == 0) return -1; //Se a projeção for 0, o raio é paralelo ou interno ao plano
    double tint = Vector(R.p0, p1).dotProduct(norma) / R.d.dotProduct(norma);
    Point Pint = R.p0 + Point(R.d.x*tint, R.d.y*tint, R.d.z*tint);
    if((Vector(p1, p2).crossProduct(Vector(p1, Pint))).dotProduct(norma) < 0) return -1;
    if((Vector(p2, p3).crossProduct(Vector(p2, Pint))).dotProduct(norma) < 0) return -1;
    if((Vector(p3, p1).crossProduct(Vector(p3, Pint))).dotProduct(norma) < 0) return -1;
    return tint;
}